leveldb_test.go

```diff
@@ -318,6 +320,179 @@ func TestIterationValidityLimits(t *testing.T) {
+		[]byte("hello world3"),
+		[]byte{}, // Yes an empty byte slice as key is valid
+		[]byte("hello world4"),
+		nil, // yes a nil key is also valid, it's interpreted as an empty byte slice
+	}
+
+	emptyKeyValue := []byte("value for empty key")
+	expectedValues := [][]byte{
+		[]byte("value for hello world0"),
+		[]byte("value for hello world1"),
+		[]byte("value for hello world2"),
+		[]byte("value for hello world3"),
+		emptyKeyValue,
+		[]byte("value for hello world4"),
+		emptyKeyValue,
+	}
+	// Populate the db with some test key-value pairs
+	for i := range keys {
+		err := db.Put(wo, keys[i], expectedValues[i])
+		if err != nil {
+			t.Errorf("Put failed: %v", err)
+		}
+	}
+
+	values, errs := db.GetMany(ro, keys)
+	if len(values) != len(errs) {
+		t.Errorf("GetMany() mismatch values len and errs len")
+	}
+	for i := range keys {
+		if errs[i] != nil {
+			t.Errorf("GetMany() failed for key[%d]: %v", i, errs[i])
+		}
+		if bytes.Compare(expectedValues[i], values[i]) != 0 {
+			t.Errorf("values[%d] is not the same as expected: %v", i, expectedValues[i])
+		}
+	}
+
+	keys2 := [][]byte{
+		[]byte("hello world0-NOT_FOUND"),
+		[]byte("hello world1"),
+		[]byte{}, // Yes an empty byte slice as key is valid
+		[]byte("hello world5-NOT_FOUND"),
+	}
+	values, errs = db.GetMany(ro, keys2)
+	if len(values) != len(errs) {
+		t.Errorf("GetMany() mismatch values len and errs len")
+	}
+	for i := range errs {
+		if errs[i] != nil {
+			t.Errorf("not expecting an error for key %d", i)
+		}
+	}
+
+	if values[0] != nil {
+		t.Errorf("Expecting non-existant key to return value of nil")
+	}
+	if bytes.Compare(expectedValues[1], values[1]) != 0 {
+		t.Errorf("values[%d] is not the same as expected: %v", 1, expectedValues[1])
+	}
+	if bytes.Compare(emptyKeyValue, values[2]) != 0 {
+		t.Errorf("values[%d] is not the same as expected value under empty key: %v", 2, emptyKeyValue)
+	}
+	if values[3] != nil {
+		t.Errorf("Expecting non-existant key to return value of nil")
+	}
+
+	values, errs = db.GetMany(ro, nil)
+	if values != nil || errs != nil {
+		t.Errorf("GetMany(): on nil slice should return nil, nil")
+	}
+
+	// Calling GetMany on a slice of N nil slices will be interpreted
+	// as N Gets of the same empty key
+	emptyKeys := make([][]byte, 10)
+	values, errs = db.GetMany(ro, emptyKeys)
+	if len(values) != len(errs) {
+		t.Errorf("GetMany() mismatch values len and errs len")
+	}
+	for i := range values {
+		if errs[i] != nil {
+			t.Errorf("GetMany() for an empty key should return something")
+		}
+		if bytes.Compare(emptyKeyValue, values[i]) != 0 {
+			t.Errorf("GetMany() for an empty key should return the value under that key")
+		}
+	}
+}
+
+// To get the benchmarks of multiple DB gets using Get() instead of the
+// GetMany() API, disable the flag like so:
+// go test -count=2 -run=^$ -bench=BenchmarkDBGets -_ldb_usegetmany=false
+var useGetMany = flag.Bool("_ldb_usegetmany", true, "By default uses uses GetMany() in favor of multiple calls of Get()")
```

Comment:
instead of this, you can do rename your `BenchmarkDBGets ` into `benchmarkDBGets(b *testing.B, useMulti bool)`

and then 
```go
func BenchmarkDBGets(b *testing.B) {
  b.Run("SingleGet", func(b *testing.B) {benchmarkDBGets(b, false)})
b.Run("MultiGet", func(b *testing.B) {benchmarkDBGets(b, true)})
}
```

Reply:
yeah I wasn't sure if that's lead to weird caching behavior if both are invoked int the same process. But can fix in the other PR that adds `PutMany()` here: https://github.com/DataDog/leveldb/pull/11

Reply:
Let's do it there so the scope of the PR is well defined ;).
What would you mean by caching behavior ? I think if we recreate new key/values before getting, nothing should be cached

Reply:
yep, no worries! The process thing is to ensure it's a clean-room, there isn't in-process buffer/caches in the test framework or below the application that could potentially pollute the results. Also not running them in one process you can clear whatever cpu/mem caches on the box before invoking it. but it's not a big deal here.